controllers/project_controller.go

```diff
@@ -0,0 +1,370 @@
+			return ctrl.Result{}, err
+		}
+	}
+
+	// Set owner reference
+	if (originalProject.OwnerReferences == nil ||
+		len(originalProject.OwnerReferences) == 0) &&
+		parentInstance != nil {
+		err := ctrl.SetControllerReference(parentInstance, originalProject, r.Scheme)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	// Update Finalizer
+	if !reflect.DeepEqual(originalProject.Finalizers, project.Finalizers) {
+		originalProject.SetFinalizers(project.Finalizers)
+	}
+
+	if err := r.Client.Update(ctx, originalProject); err != nil {
+		return ctrl.Result{Requeue: true}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// assertDeletedProject deletes a Harbor project, first ensuring its existence.
+func (r *ProjectReconciler) assertDeletedProject(ctx context.Context, log logr.Logger, harborClient *h.RESTClient,
+	project *registriesv1alpha1.Project) error {
+	_, err := harborClient.GetProjectByName(ctx, project.Name)
+	if err != nil {
+		return err
+	}
+
+	log.Info("pulling finalizers", project.Name, project.Namespace)
+	helper.PullFinalizer(project, internal.FinalizerName)
+
+	return nil
+}
+
+// assertExistingProject
+// Check Harbor projects and their components for their existence,
+// create and delete either of those to match the specification.
+func (r *ProjectReconciler) assertExistingProject(ctx context.Context, harborClient *h.RESTClient,
+	project *registriesv1alpha1.Project) error {
+	heldRepo, err := harborClient.GetProjectByName(ctx, project.Spec.Name)
+
+	if errors.Is(err, &projectapi.ErrProjectNotFound{}) {
+		_, err := harborClient.NewProject(ctx, project.Spec.Name, project.Spec.StorageLimit)
+
+		return err
+	} else if err != nil {
+		return err
+	}
+
+	return r.ensureProject(ctx, heldRepo, harborClient, project)
+}
+
+// generateProjectMetadata constructs the project metadata for a Harbor project
+func (r *ProjectReconciler) generateProjectMetadata(
+	projectMeta *registriesv1alpha1.ProjectMetadata) *model.ProjectMetadata {
+	autoScan := strconv.FormatBool(projectMeta.AutoScan)
+	enableContentTrust := strconv.FormatBool(projectMeta.EnableContentTrust)
+	preventVul := strconv.FormatBool(projectMeta.PreventVul)
+	public := strconv.FormatBool(projectMeta.Public)
+	reuseSysCVEAllowlist := strconv.FormatBool(projectMeta.ReuseSysSVEWhitelist)
+	retentionID := strconv.Itoa(projectMeta.RetentionID)
+
+	pm := model.ProjectMetadata{
+		AutoScan:             &autoScan,
+		EnableContentTrust:   &enableContentTrust,
+		PreventVul:           &preventVul,
+		Public:               public,
+		ReuseSysCveAllowlist: &reuseSysCVEAllowlist,
+		Severity:             &projectMeta.Severity,
+		RetentionID:          &retentionID,
+	}
+
+	return &pm
+}
+
+// reconcileProjectMembers reconciles the user-defined project members
+// for a project based on an existing Harbor user
+func (r *ProjectReconciler) reconcileProjectMembers(ctx context.Context, project *registriesv1alpha1.Project,
```

Comment:
This will only ensure that all users listed in the Project CR also exist in harbor. This doesn't seem to make sure deleted Users are also removed from the harbor project.